contrib/checksum/filters/http/source/checksum_filter.cc

```diff
@@ -0,0 +1,97 @@
+#include "contrib/checksum/filters/http/source/checksum_filter.h"
+
+#include <openssl/mem.h>
+
+#include "source/common/common/assert.h"
+#include "source/common/common/hex.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace ChecksumFilter {
+
+static std::vector<ChecksumFilterConfig::ChecksumMatcher> buildMatchers(
+    const envoy::extensions::filters::http::checksum::v3alpha::ChecksumConfig& proto_config) {
+  std::vector<ChecksumFilterConfig::ChecksumMatcher> matchers;
+  for (const auto& checksum : proto_config.checksums()) {
+    matchers.emplace_back(std::make_unique<Matchers::PathMatcher>(checksum.path_matcher()),
+                          Hex::decode(checksum.sha256()));
+  }
+  return matchers;
+}
+
+ChecksumFilterConfig::ChecksumFilterConfig(
+    const envoy::extensions::filters::http::checksum::v3alpha::ChecksumConfig& proto_config)
+    : matchers_(buildMatchers(proto_config)), reject_unmatched_(proto_config.reject_unmatched()) {}
+
+OptRef<const std::vector<uint8_t>> ChecksumFilterConfig::expectedChecksum(absl::string_view path) {
+  for (const auto& m : matchers_) {
+    if (m.first->match(path)) {
+      return m.second;
+    }
+  }
+  return absl::nullopt;
+}
+
+ChecksumFilter::ChecksumFilter(ChecksumFilterConfigSharedPtr config) : config_(config) {}
+
+Http::FilterHeadersStatus ChecksumFilter::decodeHeaders(Http::RequestHeaderMap& headers,
+                                                        bool /*end_stream*/) {
+  expected_checksum_ = config_->expectedChecksum(headers.Path()->value().getStringView());
+  if (!expected_checksum_.has_value() && config_->rejectUnmatched()) {
+    decoder_callbacks_->sendLocalReply(Http::Code::Forbidden, "No checksum for path", nullptr,
+                                       absl::nullopt, "no_checksum_for_path");
+    return Http::FilterHeadersStatus::StopIteration;
+  }
+  return Http::FilterHeadersStatus::Continue;
+}
+
+Http::FilterHeadersStatus ChecksumFilter::encodeHeaders(Http::ResponseHeaderMap& /*headers*/,
+                                                        bool end_stream) {
+  if (end_stream && expected_checksum_.has_value()) {
+    encoder_callbacks_->sendLocalReply(Http::Code::Forbidden, "Mismatched checksum", nullptr,
+                                       absl::nullopt, "mismatched_checksum");
+    return Http::FilterHeadersStatus::StopIteration;
+  }
```

Comment:
Please make the body content and flag more clear to tell 'expected_checksum has value but response has no body'.